cirq-core/cirq/ops/clifford_gate.py

```diff
@@ -571,3 +589,252 @@ def _circuit_diagram_info_(
+            cls._Z = cls._generate_clifford_from_known_gate(1, pauli_gates.Z)
+        return cls._Z
+
+    @property
+    def H(cls):
+        if getattr(cls, '_H', None) is None:
+            cls._H = cls._generate_clifford_from_known_gate(1, common_gates.H)
+        return cls._H
+
+    @property
+    def S(cls):
+        if getattr(cls, '_S', None) is None:
+            cls._S = cls._generate_clifford_from_known_gate(1, common_gates.S)
+        return cls._S
+
+    @property
+    def CNOT(cls):
+        if getattr(cls, '_CNOT', None) is None:
+            cls._CNOT = cls._generate_clifford_from_known_gate(2, common_gates.CNOT)
+        return cls._CNOT
+
+    @property
+    def CZ(cls):
+        if getattr(cls, '_CZ', None) is None:
+            cls._CZ = cls._generate_clifford_from_known_gate(2, common_gates.CZ)
+        return cls._CZ
+
+    @property
+    def SWAP(cls):
+        if getattr(cls, '_SWAP', None) is None:
+            cls._SWAP = cls._generate_clifford_from_known_gate(2, common_gates.SWAP)
+        return cls._SWAP
+
+
+class CommonCliffordGates(metaclass=CommonCliffordGateMetaClass):
+
+    # We need to use the lazy initialization of these common gates since they need to use
+    # cirq.sim, which can not be imported when
+    @classmethod
+    def _generate_clifford_from_known_gate(
+        cls, num_qubits: int, gate: raw_types.Gate
+    ) -> 'CliffordGate':
+        from cirq import LineQubit, sim
+
+        qubits = LineQubit.range(num_qubits)
+        t = qis.CliffordTableau(num_qubits=num_qubits)
+        args = sim.ActOnCliffordTableauArgs(
+            tableau=t, qubits=qubits, prng=np.random.RandomState(), log_of_measurement_results={}
+        )
+
+        protocols.act_on(gate, args, qubits, allow_decompose=False)
+        return CliffordGate.from_clifford_tableau(args.tableau)
+
+
+def _pad_tableau(
+    clifford_tableau: qis.CliffordTableau, num_qubits_after_padding: int, axes: List[int]
+) -> qis.CliffordTableau:
+    """Roughly, this function copies self.tabluea into the "identity" matrix."""
+    # Sanity check
+    if len(set(axes)) != clifford_tableau.n:
+        raise ValueError(
+            "Input axes of padding should match with the number of qubits in the input tableau."
+        )
+    if clifford_tableau.n > num_qubits_after_padding:
+        raise ValueError(
+            "The number of qubits in the input tableau should not be larger than "
+            "num_qubits_after_padding."
+        )
+
+    padded_tableau = qis.CliffordTableau(num_qubits_after_padding)
+    v_index = np.concatenate((np.asarray(axes), num_qubits_after_padding + np.asarray(axes)))
+
+    padded_tableau.xs[np.ix_(v_index, axes)] = clifford_tableau.xs.copy()
+    padded_tableau.zs[np.ix_(v_index, axes)] = clifford_tableau.zs.copy()
+    padded_tableau.rs[v_index] = clifford_tableau.rs.copy()
+    return padded_tableau
+
+
+@value.value_equality
+class CliffordGate(raw_types.Gate, CommonCliffordGates):
```

Comment:
Nit: We should rename this to something like `MultiQubitCliffordGate` we already have `SingleQubitCliffordGate`. This could be confusing because a user might think that `CliffordGate` is a superclass of `SingleQubitCliffordGate` which it is not.

Reply:
Renamed to `MultiQubitCliffordGate` for now. But, in my opinion, it is better to have `CliffordGate` as base class and make `SingleQubitCliffordGate` as derived class in future. (Need more work, which should not be done in this PR anyway). Because `SingleQubitCliffordGate` can reuse all `CliffordGate` functions and create more functions, like related to bloch rotation, that is unique in the Single Qubit case.

Reply:
> But, in my opinion, it is better to have CliffordGate as base class and make SingleQubitCliffordGate as derived class in future

This sounds like a good idea to investigate in a future PR.

Reply:
Yes, as discussed in the Cirq-Cynq, will make a follow-up PR to make the SingleQubitCliffordGate as the derived class of CliffordGate. So rename it back.